plugins/osdn/sdn-cni-plugin/openshift-sdn.go

```diff
@@ -0,0 +1,462 @@
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+
+	"github.com/vishvananda/netlink"
+
+	"k8s.io/kubernetes/pkg/api/resource"
+)
+
+const (
+	sdnScript   = "openshift-sdn-ovs"
+	setUpCmd    = "setup"
+	tearDownCmd = "teardown"
+	statusCmd   = "status"
+	updateCmd   = "update"
+
+	AssignMacVlanAnnotation string = "pod.network.openshift.io/assign-macvlan"
+
+	interfaceName      = "eth0"
+	ipamConfigTemplate = `{
+  "cniVersion": "0.1.0",
+  "name": "openshift-sdn",
+  "type": "openshift-sdn",
+  "ipam": {
+    "type": "host-local",
+    "subnet": "%s",
+    "routes": [
+      { "dst": "0.0.0.0/0", "gw": "%s" },
+      { "dst": "%s" }
+    ]
+  }
+}`
+)
+
+func getNetworkInfo(client *http.Client) (*api.GetNetworkResponse, error) {
+	// Try few times up to 15 seconds
+	retries := 150
+	retryInterval := 100 * time.Millisecond
+	for i := 0; i < retries; i++ {
+		network := &api.GetNetworkResponse{}
+		// HTTP client requires a URL including a domain part, but
+		// since we're talking to a unix socket the domain doesn't matter
+		if err := getUrl(client, "http://d/network", network); err == nil {
+			return network, nil
+		}
+		time.Sleep(retryInterval)
+	}
+
+	return nil, fmt.Errorf("Timed out waiting for openshift to be ready")
+}
+
+func socketDial(proto, addr string) (conn net.Conn, err error) {
+    return net.Dial("unix", api.PodInfoSocketPath)
+}
+
+func getUrl(client *http.Client, url string, object interface{}) error {
+	resp, err := client.Get(url)
+	if err != nil {
+		return fmt.Errorf("Failed to retrieve %s: %v", url, err)
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("Failed to read %s: %v", url, err)
+	}
+
+	if err := json.Unmarshal(body, object); err != nil {
+		return fmt.Errorf("Failed to read %s: %v", url, err)
+	}
+
+	return nil
+}
+
+func readPodInfo(client *http.Client, cniArgs string) (*api.GetPodResponse, error) {
+	var namespace, name string
+	for _, arg := range strings.Split(cniArgs, ";") {
+		parts := strings.Split(arg, "=")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("Invalid CNI_ARG '%s'", arg)
+		}
+		switch parts[0] {
+		case "K8S_POD_NAMESPACE":
+			namespace = strings.TrimSpace(parts[1])
+		case "K8S_POD_NAME":
+			name = strings.TrimSpace(parts[1])
+		}
+	}
+
+	if namespace == "" || name == "" {
+		return nil, fmt.Errorf("Missing pod namespace or name")
+	}
+
+	info := &api.GetPodResponse{}
+	url := fmt.Sprintf("http://d/pods/%s/%s", namespace, name)
+	if err := getUrl(client, url, info); err != nil {
+		return nil, err
+	}
+
+	return info, nil
+}
+
+// TODO: replace with k8s.io/kubernetes/pkg/util/bandwidth when rebasing to Kube 1.3
+var minRsrc = resource.MustParse("1k")
+var maxRsrc = resource.MustParse("1P")
+
+func validateBandwidthIsReasonable(rsrc *resource.Quantity) error {
+	if rsrc.Value() < minRsrc.Value() {
```

Comment:
It's basically C&P from upstream kubernetes code, and we'll replace it with that kubernetes code when the 1.3 rebase happens.  So does this mean the upstream code is also wrong?